### DISCUSSION TITLE
Problem with importing a field

### DISCUSSION THREAD

**Discussion by daredloco** [jbevain/cecil#872] [Q&A]
Hi guys, I hope someone of you can help me with that one.
I did figure out how to "overwrite" a function inside a library, but I can't get it to work with a Field that is inside the same Type as the function I did overwrite.
I know I have to "import" the field somehow, but I can't figure out how.

If i do
```
m2.Module.GetType("ModController").Fields.Add(m2.Module.ImportReference(typeof(ModController).GetField("Instance")).Resolve());
```
it would tell me "Member already attached", so I probably do something wrong. I did read the "import" part inside the documentation, but I don't quite understand it, and it doesn't talk about Fields.

Without this part, when i call target.write("filename") it would throw an exception:
```
"Member 'ModController ModController::Instance' is declared in another module and needs to be imported"
```

The code that will overwrite the old function is this one here:
```
public void LoadMods()
	{
		string[] directories = Directory.GetDirectories("DLLMods");
		foreach (string path in directories)
		{
			string[] files = Directory.GetFiles(path, "*.dll");
			if (files.Length != 0)
			{
				ModController.Instance.LoadMod(files[0], dll: true, force: false, uniqueName: false);
			}
			else
			{
				ModController.Instance.LoadMod(path, dll: false, force: false, uniqueName: false);
			}
		}
	}
```
I just don't get how to import "Instance" (which is a custom object)...

I hope anyone can help me on this one, as this seems to be the only problem by now. I did try to find more informations about that specific exception, but most "solutions" I've found were for other things than Fields.

Thanks in advance!

**Comment by ltrzesniewski** [comment]
You don't need to import fields if they're in the assembly you're modifying. They're already there. You only need to import *external* references.

The following line:

```C#
m2.Module.GetType("ModController").Fields.Add(m2.Module.ImportReference(typeof(ModController).GetField("Instance")).Resolve());
```

is trying to add a field which *already exists* in the list of fields. It also assumes the assembly you're modifying is using the same runtime as the code that's running Cecil.

Have you tried the following?

```C#
m2.Module.GetType("ModController").Fields.Single(i => i.Name == "Instance")
```

**Reply by daredloco** [reply]
> You don't need to import fields if they're in the assembly you're modifying. They're already there. You only need to import _external_ references.

I think I do missunderstand something then. What would I need to do to get the field inside the code, as my approach does not seem to work?

> Have you tried the following?
> 
> ```cs
> m2.Module.GetType("ModController").Fields.Single(i => i.Name == "Instance")
> ```
I did try it now and it won't throw an exception, so it in fact exists in the module already. I'm just not sure how I could call it from the injected code.

Could you give me a hint on how I could make it work?

Thanks for the help!

**Reply by ltrzesniewski** [reply]
I assume `ModController.Instance` is a static field, right? Then you need to:
- `ldsfld ModController::Instance` to push it on the evaluation stack:  
  `Instruction.Create(OpCodes.Ldsfld, m2.Module.GetType("ModController").Fields.Single(i => i.Name == "Instance"))`
- Push all the other arguments, from left to right
- `callvirt ModController::LoadMod`
- `pop` if `LoadMod` does not return `void`

When in doubt, you can use [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKAwGYACbUgYVIG9DSHyyMcA2UgJwFMIATAexgoAnlVIBJGAGcEEGAGMupALykYXAO4AKAJQBueo0MMS5ACykAsrtrHGDSTLmKAdABl+fS/15aARABm/Px+cKRmWPp2AL52dqYYFh5ePlosAAykQfxhsAikwBAcOnZ0BPaxBNFAA==) to tell what to emit (see the IL of the `M` method).

**Reply by daredloco** [reply]
Exactly, the ModController.Instance is a static field and is, as the name suggests, a ModController instance.

From my understanding I would need to push it to the evaluation stack inside the code that generates the new library and __not__ the code that is actually pushed to the library, right?

The LoadMod function has those parameters:
```cs
public IWorkshopItem LoadMod(string path, bool dll, bool force, bool uniqueName, string overrideName = null)
```

I did try it with:
```cs
var addInstance = Instruction.Create(OpCodes.Ldsfld, m2.Module.GetType("ModController").Fields.Single(i => i.Name == "Instance"));
var addLoadMod = Instruction.Create(OpCodes.Callvirt, m2.Module.GetType("ModController").Methods.Single(i => i.Name == "LoadMod"));

m1IL.InsertBefore(m1.Body.Instructions[0], addInstance);
m1IL.Append(Instruction.Create(OpCodes.Ldstr, ""));
m1IL.Append(Instruction.Create(OpCodes.Ldc_I4_1));
m1IL.Append(Instruction.Create(OpCodes.Ldc_I4_1));
m1IL.Append(Instruction.Create(OpCodes.Ldc_I4_1));
m1IL.Append(Instruction.Create(OpCodes.Ldstr, ""));
m1IL.Append(addLoadMod);
m1IL.Append(Instruction.Create(OpCodes.Pop));
```

But didn't have success, still the same "Member 'ModController ModController::Instance' is declared in another module and needs to be imported" exception. I did also try to add it "to the top" with InsertBefore first instruction with reversed order, but also no success.

SharpLab would give me:
```
IL_0000: ldsfld class C C::Instance
IL_0005: ldstr "foo"
IL_000a: ldc.i4.1
IL_000b: ldc.i4.1
IL_000c: ldc.i4.1
IL_000d: ldstr "other"
IL_0012: callvirt instance string C::LoadMod(string, bool, bool, bool, string)
IL_0017: pop
IL_0018: ret
```
and i did try to copy that with my approach.

I'll write the full code I have by now here:
```cs
var targetAsm = AssemblyDefinition.ReadAssembly(Path.Combine("Software Inc_Data", "Managed", "Assembly-CSharp.dll"));
var mr1 = targetAsm.MainModule.ImportReference(typeof(ModLoadTemplate).GetMethod("LoadMods"));
var targetType = targetAsm.MainModule.Types.FirstOrDefault(e => e.Name == "ModController");
var m2 = targetType.Methods.FirstOrDefault(e => e.Name == "LoadMods");
var m1 = mr1.Resolve();
var m1IL = m1.Body.GetILProcessor();



foreach (var i in m1.Body.Instructions.ToList())
{
    var ci = i;
    if (i.Operand is MethodReference)
    {
	    var mref = i.Operand as MethodReference;
	    ci = m1IL.Create(i.OpCode, targetType.Module.ImportReference(mref));
    }
    else if (i.Operand is TypeReference)
    {
	    var tref = i.Operand as TypeReference;
	    ci = m1IL.Create(i.OpCode, targetType.Module.ImportReference(tref));
    }
    if (ci != i)
    {
	    m1IL.Replace(i, ci);
    }
}


//var vars = m2.Body.Variables.ToList();
//m1.Body.Variables.Clear();
//foreach (var v in vars)
//{
//	var nv = new VariableDefinition(targetType.Module.ImportReference(v.VariableType));
//	m1.Body.Variables.Add(nv);
//}

var mLoadMod = targetType.Methods.FirstOrDefault(e => e.Name == "LoadMod");
var mLoadRef = targetType.Module.ImportReference(mLoadMod);

var addInstance = Instruction.Create(OpCodes.Ldsfld, m2.Module.GetType("ModController").Fields.Single(i => i.Name == "Instance"));
var addLoadMod = Instruction.Create(OpCodes.Callvirt, m2.Module.GetType("ModController").Methods.Single(i => i.Name == "LoadMod"));

m1IL.InsertBefore(m1.Body.Instructions[0], addInstance);
m1IL.Append(Instruction.Create(OpCodes.Ldstr, ""));
m1IL.Append(Instruction.Create(OpCodes.Ldc_I4_1));
m1IL.Append(Instruction.Create(OpCodes.Ldc_I4_1));
m1IL.Append(Instruction.Create(OpCodes.Ldc_I4_1));
m1IL.Append(Instruction.Create(OpCodes.Ldstr, ""));
m1IL.Append(addLoadMod);
m1IL.Append(Instruction.Create(OpCodes.Pop));
//m1IL.Append(Instruction.Create(OpCodes.Ret));


m2.Body = m1.Body;

targetAsm.Write("new.dll");
```
And here is the class I call with m2:
```cs
public class ModLoadTemplate
{
	public void LoadMods()
	{
		string[] directories = Directory.GetDirectories("DLLMods");
		foreach (string path in directories)
		{
			string[] files = Directory.GetFiles(path, "*.dll");
			if (files.Length != 0)
			{
				ModController.Instance.LoadMod(files[0], dll: true, force: false, uniqueName: false);
			}
			else
			{
				ModController.Instance.LoadMod(path, dll: false, force: false, uniqueName: false);
			}
		}
	}
}
```

It works if i, for example throw an exception inside the LoadMods() like throw new Exception("Test"); this would throw in fact the exception at the point LoadMods() would be called by the game. So the overwriting process actually works as expected. But trying to call or read a function/field inside the assembly that I did overwrite doesn't. Maybe I do understand something fundamentally wrong, but I'm pretty new to Cecil/IL and therefor don't know what I have to do to make it work.

Anyways, thanks alot for your help! I guess I'm on the right track due to your help and hope that I'll soon will be able to make it work.

**Reply by ltrzesniewski** [reply]
Oh, wait, you're adding code from `m2` to `m1` (two distinct modules)? In that case, yes, you need to import the stuff you reference from `m2` to `m1`.

> From my understanding I would need to push it to the evaluation stack inside the code that generates the new library and not the code that is actually pushed to the library, right?

I can't make sense of this question. There is an evaluation stack per method, which the IL instructions pop from and push to. It's a means of exchanging data between instructions.

Also, why are you inserting the first instruction with `InsertBefore` and the next one with `Append`? This seems suspicious.

In any case, good luck with this 🙂
